radix-rust/src/resolve.rs

```diff
@@ -0,0 +1,318 @@
+/// because slightly more types can implement it.
+pub trait Resolve<X: Resolvable> {
+    fn resolve(self) -> X;
+}
+
+/// The inverse trait of [`Resolve`].
+///
+/// This should be implemented instead of [`Resolve`] where possible, but
+/// [`Resolve`] should be used as bounds in arguments.
+pub trait ResolveFrom<X>: Resolvable {
+    fn resolve_from(value: X) -> Self;
+}
+
+impl<X, Y: ResolveFrom<X>> Resolve<Y> for X {
+    fn resolve(self) -> Y {
+        Y::resolve_from(self)
+    }
+}
+
+/// `Resolvable` is a marker trait, mainly to make resolution opt-in and to avoid
+/// polluting every type with a resolve method.
+///
+/// You might want to use [`resolvable_with_identity_impl`] or [`resolvable_with_try_into_impls`]
+/// to implement this trait and a reflexive or blanket impl.
+pub trait Resolvable {}
+
+#[macro_export]
+macro_rules! resolvable_with_identity_impl {
+    ($ty:ty$(,)?) => {
+        impl Resolvable for $ty {}
+
+        impl ResolveFrom<$ty> for $ty {
+            fn resolve_from(value: $ty) -> $ty {
+                value
+            }
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! resolvable_with_try_into_impls {
+    ($ty:ty$(,)?) => {
+        impl Resolvable for $ty {}
+
+        impl<T: TryInto<$ty, Error = E>, E: Debug> ResolveFrom<T> for $ty {
+            fn resolve_from(value: T) -> $ty {
+                value.try_into().unwrap_or_else(|err| {
+                    panic!(
+                        "The provided argument could not be resolved into a {}: {err:?}",
+                        core::any::type_name::<$ty>()
+                    )
+                })
+            }
+        }
+    };
+}
+
+impl<'a, X: ResolveFrom<X> + Clone> ResolveFrom<&'a X> for X {
+    fn resolve_from(value: &'a X) -> X {
+        value.clone()
+    }
+}
+
+/// This trait is intended to be used as an `impl` argument in helper methods, to accept
+/// a wider range of arguments.
+///
+/// It should only be used where it is safe to panic if the wrong argument is provided,
+/// and where performance isn't a primary concern.
+///
+/// Compared to [`Resolve`], [`LabelledResolve`] also accepts an optional resolver,
+/// which can be used to convert label/s either directly into `Self`, or into values which
+/// can be used to build up self.
+///
+/// However, unlike [`Resolve`], a reflexive [`LabelledResolve`] is only implemented for
+/// `Self`, `&Self` and various string labels. It doesn't build on top of [`TryInto`]
+/// because that causes implementation collisions with labels for types which could implement
+/// `TryFrom<&str>`.
+///
+/// ## Implementers
+/// * You should prefer to implement [`LabelledResolveFrom`] as it is easier to implement
+/// due to trait coherence rules. Sometimes you can only implement [`LabelledResolve`]
+/// however.
+/// * If requiring a labelled resolution in your bounds, prefer [`LabelledResolve`]
+/// because slightly more types can implement it.
+pub trait LabelledResolve<Y: LabelledResolvable> {
```

Comment:
It might be useful to provide a `LabelledResolve` implementation for all types that implement `Resolve` which would allow us to use this trait in all of our interfaces.

Reply:
It's not possible because of clashing blanket implementations